### PR TITLE
fix(builder): a keyboard move that the rules refuse now says why

### PR DESCRIPTION
**Pull request by mobeenabdullah**
Ledger `task:pb-keyboard-move-asks-the-nesting-question`. Files `finding:keyboard-move-bypasses-the-nesting-rule`, which **refutes** the finding this task was written from.

## The defect is bigger than the task described

The task, and this PR's first revision, said a nesting-invalid keyboard move is *"refused by the store at apply time"* and only lacked an announcement. **That is false.** Measured in a running editor:

1. insert `core/columns` (two `core/column` children)
2. select a column, focus the canvas, press `alt+ArrowLeft`
3. the column becomes a **ROOT node**, and the editor announces *"Block moved out of its container"* — a success

`core/column` declares `parent: [COLUMNS_BLOCK]`, so that is a placement a drag refuses.

It is not refused because **nothing asks**. `ops.ts` holds zero references to the nesting rule (control: the symbol resolves in `inserter.ts`), and nothing validates the document before the op applies. The pointer route is stopped only because `drop-targets` asks `blockAllowedAt` before a drop resolves.

**So a keyboard author was not merely untold — they were unstopped**, able to build a document a pointer author cannot, with nothing saying so.

## The fix

`blockAllowedAt` is asked **before** the move, in the wiring. That is the keyboard route's half of the decision the pointer route already makes — one function, two callers, so the surfaces cannot disagree about where a block may go.

`keyboard-move` is untouched. Its docblock requires that it *"answers WHERE, never WHETHER"*, and it still does; the wiring asks whether, exactly as the pointer wiring does.

Three outcomes, kept distinct:

| case | behaviour |
| --- | --- |
| the rule forbids it | **refused**, with the reason and the remedy, in the drag's own words |
| the rule permits it, the store still refuses | reported, naming **no cause** — none was established |
| a boundary (`move === null`) | silent — nowhere to go, and saying so every press is noise |

## Verified in the editor, not only in jsdom

The same steps that produced the defect now leave both columns nested and announce **"Column has to sit inside a container. Column goes inside Columns"**. Before: one column at the root, announced as a success.

## Evidence

- builder **99 files / 2740 tests**; `check-types`, `lint`, `lint:design`, comment-convention all exit 0; `fallow` `pass` over 6 files, **0 introduced**.
- The nesting test asserts `expect(editor.apply).not.toHaveBeenCalled()` — the assertion that separates **stopping** the move from narrating it, and the one whose absence let the first revision look correct.
- Break-verified against three implementations: moving the check back after `apply`, explaining the refusal but applying anyway, and announcing on a boundary. Each fails only what it should.
- The pure module's tests are mostly about the `null` — it now *permits*, so a wrong null is a document a drag would have refused.

## What I got wrong, and how

The first revision explained a refusal that, for nesting, never happens — and its test **manufactured** the branch by mocking `apply` to return `null`. Codex found that from the code; I found the same thing from the editor. Two independent measurements of one defect, and neither the task nor my own audit caught it by reading.

## Still not verified

No screen reader. The tests and the browser check assert what reaches the live region — the mechanism, not what a user hears.

### CODE REVIEW

File: packages/builder/src/keyboard-actions.tsx [hunk 1]
```diff
@@ -414,16 +438,51 @@ export function useBlockKeyboardActions({
...
+       * The reason is recovered rather than relayed, because `apply` answers
+       * `null` and carries none; and when the nesting rule would have allowed
+       * the placement, the move failed for some other reason and the sentence
+       * says only that, rather than naming a cause that was not the cause.
+       */
+      const why = refusedMoveWording(
+        editorNow.document,
+        selectedId,
+        move.to,
+        nestingSource
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/keyboard-actions.tsx]: Correct, and it is the same defect I had just found from the other end — so this is two independent measurements of one thing rather than one claim.

Yours is the reading: the op layer does not enforce nesting, so the diagnosis after `apply` was never reached for a nesting violation, and the invalid move took the SUCCESS branch. You also caught the part that made it look fine: the test mocked `apply` to return null, which manufactured the very branch that cannot occur.

Mine was in a running editor, and it is worth recording because it shows the consequence rather than the code path. I inserted `core/columns` (two `core/column` children), selected a column, focused the canvas and pressed alt+ArrowLeft. The column became a ROOT node — read back from the DOM, every marked element with its marked ancestor — and the live region said "Block moved out of its container", i.e. success. `core/column` declares `parent: [COLUMNS_BLOCK]` (column.tsx:109). So a keyboard author had just built a placement a drag refuses.

FIXED AS YOU DESCRIBE. `blockAllowedAt` is now asked BEFORE the move, in the wiring, which is the keyboard route's half of the decision `drop-targets` already makes for the pointer — one function, two callers, so the surfaces cannot disagree. `keyboard-move` is untouched and still answers WHERE and never WHETHER, as its docblock requires.

The mock is gone. The test now asserts `expect(editor.apply).not.toHaveBeenCalled()`, which is the assertion that separates STOPPING the move from narrating it. Break-verified against three implementations: moving the check back after `apply` fails it, explaining the refusal but applying anyway fails it, and announcing on a boundary fails the boundary case plus a pre-existing one.

The finding is recorded as `finding:keyboard-move-bypasses-the-nesting-rule`, linked as refuting the original — which asserted the store refuses these moves at apply time. It does not, and that sentence was the premise of both the task and this PR's first revision.

File: packages/builder/src/keyboard-actions.test.tsx [hunk 2]
```diff
@@ -355,15 +355,80 @@ describe("useBlockKeyboardActions", () => {
...
-    // The rule permitted it; the store did not.
+  it("says the move did not happen, naming no cause, when the rule permitted it", () => {
+    /*
+     * `apply` answering null means the document moved underneath the press:
+     * the nesting rule permitted the placement and the store still refused.
+     *
+     * This case asserted SILENCE until 2026-09-02. It was changed rather than
+     * added to, because silence here is the defect: a keyboard author cannot
+     * see that nothing moved, and a status that never arrives is the failure
+     * WCAG 4.1.3 describes. What has not changed is that no cause is named —
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/keyboard-actions.test.tsx]: Fixed, and you are right on the rule rather than only on the instance. The comment recorded when the assertion changed and why I changed it, which is history about the work rather than a description of the code.

It now states only the invariant: the store refusing something the nesting rule permits is reported, because a keyboard author cannot see that nothing moved, and it names no cause because none was established — a reason given there would send an author to change a container that was never the problem.

Worth admitting that this is the second time today I have written change-history into a comment in this repository, on two different files. The pull request body is where that belongs, and it is already there.

**Comment by mobeenabdullah** [comment]
@codex review

Both worked at head `ba552f656`, and your P1 changed the shape of this PR rather than patching it.

You were right that the diagnosis after `apply` could never fire for a nesting violation, and right that the test manufactured the branch by mocking `apply` to return null. I had reached the same defect from the other end an hour earlier by driving the editor: outdenting a `core/column` put it at the ROOT and announced success. Two independent measurements, and neither the task's own finding nor my audit caught it by reading — the finding actually asserts the opposite, and I have filed a refutation of it.

`blockAllowedAt` is now asked before the move, in the wiring, which is the keyboard route's half of the decision `drop-targets` already makes for the pointer. The test asserts `apply` is not reached, which is the assertion whose absence let the first revision look right.

What I would still like attacked: the three outcomes have to stay distinct — refuse with a reason, report with no cause, and stay silent at a boundary — and the middle one is the one I am least sure of. If the store refuses something the rule permits, I announce that the block did not move and name nothing. If you think that sentence is still too confident, or that there is a path reaching it that should have been a boundary, that is the finding I want.

Unchanged and worth repeating: no screen reader here. The assertions cover what reaches the live region, not what is heard.

File: packages/builder/src/keyboard-actions.tsx [hunk 3]
```diff
@@ -404,26 +428,61 @@ export function useBlockKeyboardActions({
       // be an error message for pressing a key that did nothing.
       if (move === null) return;
 
+      /*
+       * ASKED BEFORE THE MOVE, which is the whole of this fix.
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/keyboard-actions.tsx]: Fixed, and the rule is the one I keep breaking rather than a one-off. Three of my comments recorded how the code CHANGED — 'the whole of this fix', 'before this', 'the gap this closes' — and one more in `move-refusal`'s module docblock said 'until this module was called before the move rather than after it'.

All four now state only what holds: nesting permission is asked before `apply` because the store does not enforce it, and this is the keyboard route's half of the decision `drag-targets` makes for the pointer.

This is the third time today you or CodeRabbit have caught me writing change-history into a comment, across three different files in three different pull requests. The pattern is that I reach for it exactly when I have just fixed something and the contrast feels like the clearest explanation — which is the moment the rule is designed for, because the contrast is only clear to someone who saw both states. The PR body is where it belongs and it is already there.

File: packages/builder/src/keyboard-actions.tsx [hunk 4]
```diff
@@ -404,26 +428,61 @@ export function useBlockKeyboardActions({
...
+       * cannot, and nothing says so.
+       */
+      const refusal = nestingRefusalForMove(
+        editorNow.document,
+        selectedId,
+        move.to,
+        nestingSource
+      );
+      if (refusal !== null) {
+        announce(refusalAnnouncement(refusal));
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/builder/src/keyboard-actions.tsx:448]: This is the more valuable of the two and I would not have found it. `canvas.tsx` documented the OPPOSITE of the new behaviour on four counts: that the keyboard path never asks the nesting question, that a move is refused only by the store, that `keyboard-actions` returns silently, and that a test pins that silence. All four are now false.

A note describing a relationship the change inverted is worse than no note, because the next reader has no reason to distrust it — and this one sits at the exact element whose refusal is now spoken elsewhere, so it is the first place someone would look.

It now names the spoken counterpart, records that `keyboardMovePosition` stays purely positional so the question is asked by the wiring on both routes of one function, and keeps the reason announcing on the pointer gesture would still be wrong: it would put the sentence on a gesture, spoken to someone who is not the one dragging.
